allennlp/nn/beam_search.py

```diff
@@ -25,6 +27,365 @@
+
+
+@Sampler.register("deterministic")
+class DeterministicSampler(Sampler):
+    """
+    A `Sampler` that just deterministically returns the `k` nodes or beams with highest
+    log probability.
+    """
+
+    @overrides
+    def sample_nodes(
+        self, log_probs: torch.Tensor, per_node_beam_size: int, state: StateType
+    ) -> Tuple[torch.Tensor, torch.Tensor, StateType]:
+        selected_log_probs, selected_indices = torch.topk(log_probs, per_node_beam_size, dim=-1)
+        return selected_log_probs, selected_indices, {}
+
+
+@Sampler.register("multinomial")
+class MultinomialSampler(Sampler):
+    """
+    Represents a sampler to choose values from a multinomial distribution.
+    Registered as a `Sampler` with name "multinomial".
+    `log_probs` is a tensor of normalized log-probabilities to be selected from.
+    `temperature` modules the probabilities of the selected tokens. A `temperature` below
+    1.0 produces a sharper probability distribution and a `temperature` above 1.0 produces
+    a flatter probability distribution.
+    `with_replacement` is `False` if sampled instances must be repeated in the sampled set.
+    Default is `True`.
+    """
+
+    def __init__(
+        self,
+        temperature: float = 1.0,
+        with_replacement: bool = False,
+    ) -> None:
+        self.temperature = temperature
+        self.filter_val = min_value_of_dtype(torch.float)
+        self.with_replacement = with_replacement
+
+    @overrides
+    def sample_nodes(
+        self, log_probs: torch.Tensor, per_node_beam_size: int, state: StateType
+    ) -> Tuple[torch.Tensor, torch.Tensor, StateType]:
+        if self.temperature != 1.0:
+            _probabilities = torch.nn.functional.softmax(log_probs / self.temperature, dim=-1)
+        else:
+            _probabilities = log_probs.exp()
+
+        selected_indices = torch.multinomial(
+            _probabilities, per_node_beam_size, replacement=self.with_replacement
+        )
+
+        return torch.gather(log_probs, 1, selected_indices), selected_indices, state
+
+
+@Sampler.register("top-k")
+class TopKSampler(Sampler):
+    """
+    Represents a `Sampler` which redistributes the probability mass function among
+    the top `k` choices then selects from that subset
+    `k` is the number of highest-probability options that the returned choice will be selected from.
+    `temperature` modules the probabilities of the selected tokens. A `temperature` below 1.0 produces a
+    sharper probability distribution and a `temperature` above 1.0 produces a flatter probability
+    distribution.
+    Registered as a `Sampler` with name "top-k".
+    """
+
+    def __init__(
+        self,
+        k: int = 1,
+        temperature: float = 1.0,
+        filter_val: float = -float("inf"),
+        with_replacement: bool = False,
+    ):
+        if not isinstance(k, int) or k < 1:
+            raise ValueError("k must be a positive integer")
```

Comment:
Just a minor nit. We generally don't check types since we have type hints. I think checking the value is sufficient.

```suggestion
        if k < 1:
            raise ValueError("k must be a positive integer")
```

Reply:
Actually, with the suggested change below, I don't think we need this.